src/Shared/FileMatcher.cs

```diff
@@ -110,21 +110,51 @@ internal FileMatcher(IFileSystem fileSystem, GetFileSystemEntries getFileSystemE
 
             _getFileSystemEntries = getFileSystemDirectoryEntriesCache == null
                 ? getFileSystemEntries
-                : (type, path, pattern, directory, projectDirectory) =>
+                : (type, path, pattern, directory, stripProjectDirectory) =>
                 {
-                    // Cache only directories, for files we won't hit the cache because the file name patterns tend to be unique
-                    if (type == FileSystemEntity.Directories)
+                    if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
                     {
-                        return getFileSystemDirectoryEntriesCache.GetOrAdd(
-                            $"{path};{pattern ?? "*"}",
-                            s => getFileSystemEntries(
-                                type,
-                                path,
-                                pattern,
-                                directory,
-                                projectDirectory).ToArray());
+                        // New behavior:
+                        // Always hit the filesystem with "*" pattern, cache the results, and do the filtering here.
+                        string cacheKey = type switch
+                        {
+                            FileSystemEntity.Files => "F",
+                            FileSystemEntity.Directories => "D",
+                            FileSystemEntity.FilesAndDirectories => "A",
+                            _ => throw new NotImplementedException()
+                        } + ";" + path;
```

Comment:
Nit:
Use $" to prevent intermediate allocation?

Reply:
The compiler generates a call to `String.Concat(string, string, string)` in both cases: https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMCADJgLIIAUF1AzggDTZWbNwCUhxUkJgB2bgkwBqTACIA3DKnc4cwSQC+qNWU7UacRl1Ycmy/kRRDhYgCQyCrdXIdx187ZpTqgA

In cases where most of the expression is literal, I find `$"` useful, e.g. `$"Hello {name}, how are you?"`. Otherwise, when there's only a little literal string and there's no perf impact, I prefer the good old `+`.